File: 1501~2000/1759. Count Number of Homogenous Substrings/main.py
"""
Source: https://rb.gy/jm8v9
Date: 2023/7/26
Skill:
Ref:
Runtime: 149 ms, faster than 70.83%
Memory Usage: 17.57 MB, less than 15.63%
Time complexity:
Space complexity:
Constraints:

"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def countHomogenous(self, s: str) -> int:
        MOD = 10 ** 9 + 7
        res, cur_char, cur_num, sz = 1, s[0], 1, len(s)
        for i in range(1, sz):
            if s[i] == cur_char:
                cur_num += 1
                res += cur_num
            else:
                cur_char = s[i]
                cur_num = 1
                res += cur_num
        return res % MOD


if __name__ == "__main__":
    s = Solution()

File: 1501~2000/1759. Count Number of Homogenous Substrings/main.cpp
/**
 * Source: ibit.ly/z4JHb
 * Date: 2023/11/9
 * Skill:
 * Runtime: 20 ms, faster than 93.70% of C++ online submissions
 * Memory Usage: 11.94 MB, less than 85.65% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */


#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>


using namespace std;

#define ll long long

typedef pair<int, int> pairs;

const int MX = 2 * 1e5 + 2;

class Solution {
public:
    int countHomogenous(string s) {
        ll mod = 1e9+7, res = 0;
        int idx = -1, sz = s.length();
        while (++idx < sz) {
            ll num = 1;
            while (idx + 1 < sz && s[idx] == s[idx + 1]) {
                idx++;
                num++;
            }
            res = (res + (1 + num) * num / 2) % mod;
        }
        return res;
    }
};


int main() {
    Solution s;
    vector<vector<int>> arr{{4,8}, {2,8}};
    vector<int> nums{2,3,5};
}